src/Chapter3Task.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE InstanceSigs #-}
+
+module Chapter3Task where
+
+newtype Health = MkHealth {unHealth :: Integer} deriving (Show, Eq)
+
+newtype Attack = MkAttack {unAttack :: Integer} deriving (Show, Eq)
+
+newtype Defense = MkDefense {unDefense :: Integer} deriving (Show, Eq)
+
+data KnightAction
+  = KnightAttack 
+  | DrinkPotion 
+  | CastSpell 
+  deriving (Show, Eq)
+
+data MonsterAction
+  = MonsterAttack
+  | RunAway
+  deriving (Show, Eq)
+
+data Action
+  = KnightAction KnightAction
+  | MonsterAction MonsterAction
+  deriving (Show, Eq)
+
+data Knight = Knight {
+  knightHealth :: Health,
+  knightAttack :: Attack,
+  knightDefense :: Defense,
+  knightActions :: [Action] -- is there a way to limit to KnightActions only here? Can this lead to runtime errors?
```

Comment:
Yes, you can use the `KnightAction` type here directly instead of `Action`. As you correctly noticed, this will lead to a safer design.